osu.Game.Tests/Rulesets/TestSceneRulesetSkinProvidingContainer.cs

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using NUnit.Framework;
+using osu.Framework.Allocation;
+using osu.Framework.Audio.Sample;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.OpenGL.Textures;
+using osu.Framework.Graphics.Textures;
+using osu.Game.Audio;
+using osu.Game.Rulesets;
+using osu.Game.Skinning;
+using osu.Game.Tests.Testing;
+using osu.Game.Tests.Visual;
+
+namespace osu.Game.Tests.Rulesets
+{
+    public class TestSceneRulesetSkinProvidingContainer : OsuTestScene
+    {
+        [Resolved]
+        private SkinManager skins { get; set; }
+
+        private SkinRequester requester;
+
+        protected override Ruleset CreateRuleset() => new TestSceneRulesetDependencies.TestRuleset();
+
+        [Test]
+        public void TestEarlyAddedSkinRequester()
+        {
+            Texture textureOnLoad = null;
+
+            AddStep("set skin", () => skins.CurrentSkinInfo.Value = DefaultLegacySkin.Info);
+
+            AddStep("setup provider", () =>
+            {
+                var rulesetSkinProvider = new RulesetSkinProvidingContainer(Ruleset.Value.CreateInstance(), Beatmap.Value.Beatmap, Beatmap.Value.Skin);
+
+                rulesetSkinProvider.Add(requester = new SkinRequester());
+
+                requester.OnLoadAsync += () => textureOnLoad = requester.GetTexture("hitcircle");
```

Comment:
What's the reasoning for doing this event binding flow? Why not just do it inside of the class itself?

Reply:
Also where is "hitcircle" coming from? Shouldn't you be adding a resource fetch local and unique to this test scene so we can be sure it's actually doing as expected?

Reply:
> What's the reasoning for doing this event binding flow? Why not just do it inside of the class itself?

I was gonna do that at first of course, but felt weird for a class to do logic unrelated to it, given that there'll be more test cases in this test scene, so I've decided to go the event binding way instead to keep things inside the test case.

> Also where is "hitcircle" coming from? Shouldn't you be adding a resource fetch local and unique to this test scene so we can be sure it's actually doing as expected?

The step above this one sets the current `SkinManager` skin to default legacy skin, so the "hitcircle" would come from it, if the skin provider is working correctly. Not sure what you mean by "adding a resource fetch local", but I can change this to fetch from ruleset's resources in #13441, to make sure it's working as expected.

Reply:
Unless you mean actually making a `RulesetSkinProvidingContainer` subclass that would have a custom `ISkin` added to it for the testing, which sounds a bit too much?

Reply:
At very least you need to add a local source to ensure we're getting the expectation. You can't just use a global resource like "hitcircle" because to someone reading these tests it makes zero sense as to where this expectation comes from. Explicitly show the source in the test, please.

Reply:
```diff
index b6800e40e4..0f64db6734 100644
--- a/osu.Game.Tests/Rulesets/TestSceneRulesetSkinProvidingContainer.cs
+++ b/osu.Game.Tests/Rulesets/TestSceneRulesetSkinProvidingContainer.cs
@@ -2,6 +2,7 @@
 // See the LICENCE file in the repository root for full licence text.

 using System;
+using JetBrains.Annotations;
 using NUnit.Framework;
 using osu.Framework.Allocation;
 using osu.Framework.Audio.Sample;
@@ -10,6 +11,7 @@
 using osu.Framework.Graphics.OpenGL.Textures;
 using osu.Framework.Graphics.Textures;
 using osu.Game.Audio;
+using osu.Game.Beatmaps;
 using osu.Game.Rulesets;
 using osu.Game.Skinning;
 using osu.Game.Tests.Testing;
@@ -35,11 +37,11 @@ public void TestEarlyAddedSkinRequester()

             AddStep("setup provider", () =>
             {
-                var rulesetSkinProvider = new RulesetSkinProvidingContainer(Ruleset.Value.CreateInstance(), Beatmap.Value.Beatmap, Beatmap.Value.Skin);
+                var rulesetSkinProvider = new TestSkinProvidingContainer(Ruleset.Value.CreateInstance(), Beatmap.Value.Beatmap, Beatmap.Value.Skin);

                 rulesetSkinProvider.Add(requester = new SkinRequester());

-                requester.OnLoadAsync += () => textureOnLoad = requester.GetTexture("hitcircle");
+                requester.OnLoadAsync += () => textureOnLoad = requester.GetTexture(TestSkinProvider.TEXTURE_NAME);

                 Child = rulesetSkinProvider;
             });
@@ -69,5 +71,33 @@ private void load(ISkinSource skin)

             public IBindable<TValue> GetConfig<TLookup, TValue>(TLookup lookup) => skin.GetConfig<TLookup, TValue>(lookup);
         }
+
+        private class TestSkinProvidingContainer : RulesetSkinProvidingContainer
+        {
+            public TestSkinProvidingContainer(Ruleset ruleset, IBeatmap beatmap, [CanBeNull] ISkin beatmapSkin)
+                : base(ruleset, beatmap, beatmapSkin)
+            {
+            }
+
+            protected override void UpdateSkinSources()
+            {
+                base.UpdateSkinSources();
+
+                SkinSources.Insert(0, new TestSkinProvider());
+            }
+        }
+
+        private class TestSkinProvider : ISkin
+        {
+            public const string TEXTURE_NAME = "some-texture";
+
+            public Drawable GetDrawableComponent(ISkinComponent component) => throw new NotImplementedException();
+
+            public Texture GetTexture(string componentName, WrapMode wrapModeS, WrapMode wrapModeT) => componentName == TEXTURE_NAME ? Texture.WhitePixel :
null;
+
+            public ISample GetSample(ISampleInfo sampleInfo) => throw new NotImplementedException();
+
+            public IBindable<TValue> GetConfig<TLookup, TValue>(TLookup lookup) => throw new NotImplementedException();
+        }
     }
 }
```
Like this? I'm completely unsure about the addition of `TestSkinProvidingContainer` for doing that, but that'll be needed.

Reply:
Am I missing something? Can't you just cache a (parent) source on the test scene rather than doing local insertion? Wouldn't that match usage closer?

But yes, this looks more explicit at least.

Reply:
Right... been thinking like how `RulesetSkinProvidingContainer` behaved previously (only cached `SkinManager`), that would work.